gateway/upstreamer/push/upstreamer.go

```diff
@@ -66,44 +66,56 @@ func (c *Upstreamer) Upstream(req *http.Request) (string, float64) {
 		n1, n2 = 0, 1
 
 	default:
-		n1, n2 = pick(l)
+		c.config.lock.Lock()
+		n1, n2 = pick(c.config.randomizer, l)
```

Comment:
Let pick() lock. This is weird (edit: as stated above, this lock should not be here)